ucp/_libs/ucx_api.pyx

```diff
@@ -372,17 +378,227 @@ def _ucx_worker_handle_finalizer(
     ucp_worker_destroy(handle)
 
 
+IF CY_UCP_AM_SUPPORTED:
+    cdef void _am_recv_completed_callback(
+        void *request,
+        ucs_status_t status,
+        size_t length,
+        void *user_data
+    ):
+        cdef bytearray buf
+        cdef UCXRequest req
+        cdef dict req_info
+        cdef str name, ucx_status_msg, msg
+        cdef set inflight_msgs
+        cdef tuple cb_args
+        cdef dict cb_kwargs
+
+        logger.debug(
+            "_am_recv_completed_callback status %d len %d buf %s" % (
+                status, length, hex(int(<uintptr_t>user_data))
+            )
+        )
+
+        assert user_data != NULL
+
+        with log_errors():
+            req = UCXRequest(<uintptr_t><void*> request)
+            assert not req.closed()
+            req_info = <dict>req._handle.info
+            req_info["status"] = "finished"
+
+            if "cb_func" not in req_info:
+                # This callback function was called before _am_recv_callback() returned
+                return
+
+            exception = None
+            if status == UCS_ERR_CANCELED:
+                name = req_info["name"]
+                msg = "<%s>: " % name
+                exception = UCXCanceled(msg)
+            elif status != UCS_OK:
+                name = req_info["name"]
+                ucx_status_msg = ucs_status_string(status).decode("utf-8")
+                msg = "<%s>: %s" % (name, ucx_status_msg)
+                exception = UCXError(msg)
+            try:
+                inflight_msgs = req_info["inflight_msgs"]
+                inflight_msgs.discard(req)
+                cb_func = req_info["cb_func"]
+                if cb_func is not None:
+                    cb_args = req_info["cb_args"]
+                    if cb_args is None:
+                        cb_args = ()
+                    cb_kwargs = req_info["cb_kwargs"]
+                    if cb_kwargs is None:
+                        cb_kwargs = {}
+                    cb_func(cb_args[0], exception, **cb_kwargs)
+            finally:
+                req.close()
+
+    cdef ucs_status_t _am_recv_callback(
```

Comment:
We may want to add some directives here. In particular `boundscheck` to avoid checking if indexing `buf[0]` will raise an error. Disabling some of the other checks may make sense as well

https://github.com/rapidsai/ucx-py/blob/0fca081f960aa0372e1d2c6414c8c74ecc3a9b11/ucp/_libs/arr.pyx#L203-L206

Reply:
This is done now.